### DISCUSSION TITLE
Enable the devtools per iframe

### DISCUSSION THREAD

**Discussion by hwillson** [apollographql/apollo-client-devtools#380] [Ideas]
Migrated from: https://github.com/apollographql/apollo-client-devtools/issues/120

With pages that contain iframes, such as codesandbox, we should be able to access the Apollo Client instance for an embedded iframe.

**Comment by patjakubik** [comment]
Yes please. This is big deal when building embedded Shopify apps.

**Reply by donnellyjoe** [reply]
This would also be great for aiding development of UI components within storybook 👍

**Comment by boehlkers** [comment]
Hi @hwillson - I’ve come across this as a  requirement in a project I’m working on and addressing it would be a huge QoL improvement. Would you be open to a PR that implements this (by checking for the existence of any `window.__APOLLO_CLIENT__` instance in any embedded iframe on the page on extension startup)?

The only concern I could see is in the case where multiple instances appear on the same page within different iframes. But as long as we prioritize the top-level `window.__APOLLO_CLIENT__` instance there should be no change to any existing functioning use cases.

**Reply by karthikiyengar** [reply]
> by checking for the existence of any window.__APOLLO_CLIENT__ instance in any embedded iframe on the page on extension startup

I'm not sure that would work for iframes where the Same-Origin Policy is not honoured.

(Correction: all_frames should kinda do it)

---
It seems like React DevTools [gets around this](https://github.com/facebook/react-devtools/pull/440/files) by packaging a standalone Electron version that allows connections using WebSockets. An approach like this would help us support Chrome extensions as well, as requested here: https://github.com/apollographql/apollo-client-devtools/discussions/382

I haven't explored how much of an effort it would be to package this as an Electron app, has this been considered?

**Reply by karthikiyengar** [reply]
I was playing around with what's possible and it looks like we can achieve this by using `allFrames: true` on the content script. We might have to rework the messaging a bit to also include an iframe identifier, potentially allowing us to select the window we want to use DevTools with. Thoughts, @hwillson ?

**Comment by karthikiyengar** [comment]
[Draft PR here](https://github.com/apollographql/apollo-client-devtools/pull/951)

**Comment by torressam333** [comment]
Did this ever make it to prod? Are we able to invoke dev tools as an executable script?

**Reply by jerelmiller** [reply]
@torressam333 not yet! It's on the short list of next upcoming things to get in place. We've been putting some effort recently to ensure the devtools is more stable first as we've had a lot of reports that it just refuses to work at all. Once we feel good about that, that should free us up to add these features. Hopefully sooner than later!

**Comment by sagarpanchal** [comment]
Any update on this?

**Reply by phryneas** [reply]
Not for the browser extension so far, but you can use the VSCode version of the DevTools with iFrames: https://www.apollographql.com/docs/react/development-testing/developer-tooling#apollo-client-devtools-in-vs-code
